### DISCUSSION TITLE
クエリー構文として空文字列などを与えるとエラーになる

### DISCUSSION THREAD

**Discussion by roodni** [pgroonga/pgroonga#377] [Q&A]
OS: macOS Ventura 13.4.1 (c)
Postgresql Version: 14.9
PGroonga Version: 3.1.3
libgroonga Version: 13.0.9

クエリー構文が要求される箇所、例えば`pgroonga_query_extract_keywords`関数や`&@~`演算子に空文字列を与えると、パース失敗のエラーが発生します。

```
# SELECT pgroonga_query_extract_keywords('');
ERROR:  pgroonga: [query-extract-keywords] failed to parse expression: <>: Syntax error: <||>

# SELECT f FROM UNNEST(ARRAY['a','b']) AS f WHERE f &@~ '';
ERROR:  pgroonga: [sequential-search][query] failed to parse expression: <>: Syntax error: <||>
```

空文字列のほかに、半角空白だけの文字列でも同様のエラーが発生します。
```
# SELECT pgroonga_query_extract_keywords(' ');
ERROR:  pgroonga: [query-extract-keywords] failed to parse expression: < >: Syntax error: < ||>
```
全角空白や改行文字でも同様です。

`pgroonga_query_escape`関数を噛ませるのも試しましたが、だめでした。
```
# SELECT pgroonga_query_extract_keywords(pgroonga_query_escape(''));
ERROR:  pgroonga: [query-extract-keywords] failed to parse expression: <>: Syntax error: <||>
```

このことは、「ユーザがwebフォームの検索窓に入力した内容をクエリー構文として解釈し検索する」ような利用をPGroongaは想定していない、ということでしょうか。
実際 https://pgroonga.github.io/ja/users/#topstudio で述べられた利用例では、検索文字列を自力でパースして簡易的なAND検索を実現しているようです。しかし、クエリー構文を利用できるならば、私はそれに頼りたいと考えています。

**Comment by kou** [comment]
想定しているんですが、`f &@~ ''`はどういう挙動になるのがいいですかねぇ。
0ヒットでいいのかしら。

`pgroonga_query_extract_keywords('')`は`[]`で大丈夫そう。

**Reply by roodni** [reply]
ご回答ありがとうございます。
仕様ではなくバグでしたか。修正されることを期待して良いようで嬉しいです。

`f &@ ''`が0ヒットになることを踏まえると、`f &@~ ''`も0ヒットになるが自然な挙動に見えますね。

**Reply by kou** [reply]
それで自然かやってみましょうか。

* #378
* #379
